### PR TITLE
Introduce filters to modify/inspect the blocks returned by `do_blocks()`

### PR DESCRIPTION
**Pull request by luisherranz**
The `do_blocks()` function provides a way to process blocks including block parsing and their subsequent rendering. Currently, it is impossible for plugins (including Gutenberg) to inspect those values or customize that logic.  

A couple of filters should be introduced that allow to modify or inspect the blocks before they are included in the final response:  

- The first one receives the array of parsed blocks returned by the `parse_blocks` function. This filter provides an easy way to inspect or even modify the blocks as a whole, and also gives the opportunity to know when a call to `do_blocks` has started.
- The second one receives the final HTML rendered returned by the recursive call to `render_block`. This filter provides an easy way to inspect or even modify the final rendered output as whole, and also gives the opportunity to know when a call to `do_blocks` is about to finish.

With the combination of these two filters, it's also possible to ignore nested calls of `do_blocks` and avoid processing the HTML of blocks more than once, something that is not possible today using `render_block` alone.  

---

Trac ticket: https://core.trac.wordpress.org/ticket/59131

#### Questions

- ~~Do I need to add tests for these filters?~~ (done)
- Are the `do_blocks_pre/post_render` filter names ok?
- @dmsnell: is this what you had in mind, or is there something missing?

### CODE REVIEW

File: src/wp-includes/blocks.php [hunk 1]
```diff
@@ -1166,12 +1166,32 @@ function parse_blocks( $content ) {
  */
 function do_blocks( $content ) {
 	$blocks = parse_blocks( $content );
+
+	/**
+	 * Filter to allow plugins to inspect or change the array of parsed block
```
Comment (dmsnell) [inline src/wp-includes/blocks.php]: a bit of a pedantic note, but this is a tree of parsed blocks. we are probably inconsistent in Core in our use of the terms, but it could help to subtly communicate that there's nesting and recursion involved - it's not a flat array

**Review by dmsnell — COMMENTED** [review]
love the names and the argument list. I wondered if we would want to pass additional args: the original HTML content, the post id, etc… but in each case I don't think we should send more. most of the ids, if available, are available through globals. the original HTML content isn't provided, but there's no clear way to associate specific parts of it with the parsed blocks and so I think it's best to leave that out unless we find a clear need for it.

in other words, looks very nice.

**Comment by dmsnell** [comment]
> Do I need to add tests for these filters?

This would be nice. Check in `phpunit/tests/blocks/render.php` though I could easily see a valid use case for adding a new `renderHooks.php` test file. Something easy like filtering out image blocks or paragraphs containing `Wordpress` could suffice; just enough to show that the filter is called and works. Same for the `_post_render` filter.

**Comment by luisherranz** [comment]
Thanks, Dennis!

I've added the tests and improved the comments. Let me know if there's something else that can be done.

File: tests/phpunit/tests/blocks/renderHooks.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	/**
+	 * @ticket 59131
+	 */
+	public function test_do_blocks_pre_render_filter() {
+		$block_content = '
+			<!-- wp:core/paragraph --><p>Hello</p><!-- /wp:core/paragraph -->
+			<!-- wp:core/paragraph --><p>WordPress!</p><!-- /wp:core/paragraph -->
+		';
+
+		add_filter( 'do_blocks_pre_render', array( $this, 'do_blocks_pre_render_filter' ) );
```
Comment (dmsnell) [inline tests/phpunit/tests/blocks/renderHooks.php]: it would be nice if these had names relating to what they are doing to the test rather than names that explain where they go. that is, if I see `array( $this, 'remove_blocks_not_containing_wordpress' )` then I don't have to jump around looking for the filter function and then work backwards. same for the `post_` filter.

this is not a blocking point, but a nice to have

File: src/wp-includes/blocks.php [hunk 3]
```diff
@@ -1166,12 +1166,33 @@ function parse_blocks( $content ) {
...
+	/**
+	 * Filter to allow plugins to inspect or manipulate the final HTML output.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string $output String of rendered HTML.
+	 */
+	$output = apply_filters( 'do_blocks_post_render', $output );
+
 	// If there are blocks in this content, we shouldn't run wpautop() on it later.
```
Comment (gziolo) [inline src/wp-includes/blocks.php:1560]: The interesting part of introducing these filters is that we could even consider moving the logic below as a separate filter applied with `do_blocks_post_render`.

File: tests/phpunit/tests/blocks/renderHooks.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+		$result = do_blocks( $block_content );
+
+		$this->assertSame( '<p>Hello</p>', trim( $result ) );
+
+		remove_filter( 'do_blocks_pre_render', array( $this, 'do_blocks_pre_render_filter' ) );
```
Comment (gziolo) [inline tests/phpunit/tests/blocks/renderHooks.php]: Minor thing, but the filter could be moved before the assertion so it gets executed even when the assertion fails. This increases the chances the changes from the filter won't leak into other test cases causing random issues.

File: src/wp-includes/blocks.php (hunk 1 above)
Comment (luisherranz) [reply src/wp-includes/blocks.php]: Done in https://github.com/WordPress/wordpress-develop/pull/5023/commits/c5cfbc393562991f969e14d21ce9c76dc4a03f92.

File: tests/phpunit/tests/blocks/renderHooks.php (hunk 2 above)
Comment (gziolo) [reply tests/phpunit/tests/blocks/renderHooks.php]: It's also possible inline the filter using local functions so it's colocated in case it won't be used in other places.

See the example in the codebase:

https://github.com/WordPress/wordpress-develop/blob/2acd0aca7845feda1c4ca4dd708ca4e8a72c422b/tests/phpunit/tests/blocks/register.php#L930C1-L937

File: tests/phpunit/tests/blocks/renderHooks.php (hunk 4 above)
Comment (luisherranz) [reply tests/phpunit/tests/blocks/renderHooks.php]: Makes sense, thanks. Done in https://github.com/WordPress/wordpress-develop/pull/5023/commits/69cbad7d5a3176169712b4396d08ec6a723339ad.

File: tests/phpunit/tests/blocks/renderHooks.php (hunk 2 above)
Comment (luisherranz) [reply tests/phpunit/tests/blocks/renderHooks.php]: Makes sense, thanks. Done in https://github.com/WordPress/wordpress-develop/pull/5023/commits/230a9ade032f1d2f3db85951c221903ef0445428.

File: src/wp-includes/blocks.php (hunk 3 above)
Comment (luisherranz) [reply src/wp-includes/blocks.php:1560]: `do_blocks_post_render` only gets `$output` (HTML), but not `$content` (parsed blocks).

So can that logic be applied using the `do_blocks_pre_render` filter, or maybe it makes sense to pass `$content` also to the `do_blocks_post_render` filter?

**Comment by gziolo** [comment]
> Do I need to add tests for these filters?

Always, but I see you have it covered 😄 

> Are the do_blocks_pre/post_render filter names ok?

I don't think there is a hard rule for it, but the shorter the name, the better, as long as it expresses the intent. I usually look at existing names when trying to decide. Here I did a quick search at https://developer.wordpress.org/?s=pre_&post_type%5B%5D=wp-parser-hook:

Screenshot 2023-08-18 at 10 10 37
Screenshot 2023-08-18 at 10 10 51

There is [`pre_render_block`](https://developer.wordpress.org/reference/hooks/pre_render_block/) in `render_block` called inside `do_blocks`, so you could as well follow and go with `pre_do_blocks` and `post_do_blocks`. I don't have strong preferences as the current naming used is also good.

File: tests/phpunit/tests/blocks/renderHooks.php (hunk 2 above)
Comment (luisherranz) [reply tests/phpunit/tests/blocks/renderHooks.php]: > It's also possible inline the filter using local functions so it's colocated in case it won't be used in other places.

Oh, ok. I didn't see that comment.

Done in https://github.com/WordPress/wordpress-develop/pull/5023/commits/5abfc9c4a2c81b82bdb489211db2493517e5b26b.

**Comment by luisherranz** [comment]
> You could as well follow and go with pre_do_blocks and post_do_blocks.

I went with the extra "render" word because it leaves the door open for a future `do_blocks_pre_parse` filter where you can modify the `$content` before it gets parsed, but I'm not sure if that one makes sense, though.

File: src/wp-includes/blocks.php (hunk 3 above)
Comment (gziolo) [reply src/wp-includes/blocks.php:1560]: > do_blocks_post_render only gets $output (HTML), but not $content (parsed blocks).

I missed that.

In https://github.com/WordPress/wordpress-develop/pull/5023#pullrequestreview-1582962695 @dmsnell mentioned:

>  I wondered if we would want to pass additional args: the original HTML content, the post id, etc

That might make sense to pass the `$content` with both hooks. I often see additional parameters passed with the hook that help with processing in plugins.

`$post_id` isn't explicitly used in the function, and you can always read it with the `get_post` call. I don't think it is needed here.

**Comment by gziolo** [comment]
> > You could as well follow and go with pre_do_blocks and post_do_blocks.
> 
> I went with the extra "render" word because it leaves the door open for a future `do_blocks_pre_parse` filter where you can modify the `$content` before it gets parsed, but I'm not sure if that one makes sense, though.

The name is fine, like I said before. We wouldn't need to introduce another filter unless the body changes drastically. You only need to pass the `$content` to achieve the same result:

```php
$blocks = apply_filters( 'do_blocks_pre_render', $blocks, $content );

// usage
add_filter( 'do_blocks_pre_render', function( $content ) {
    return parse_blocks( my_plugin_modify_content( $content );
}, 10, 1 ); 
```

File: src/wp-includes/blocks.php (hunk 3 above)
Comment (luisherranz) [reply src/wp-includes/blocks.php:1560]: > That might make sense to pass the $content with both hooks.

Ok, let's see. Right now the filters are:

- `$blocks = apply_filters( 'do_blocks_pre_render', $blocks );`
- `$output = apply_filters( 'do_blocks_post_render', $output );`

And we have these variables:

- `$content`: Post content (unprocessed).
- `$blocks`: Array of parsed blocks.
- `$output`: Processed post content

It makes sense to pass `$content` to `pre`, but which variables do you think we should pass to `post`? Only `$content`?

**Comment by luisherranz** [comment]
Okay, so `parse_blocks` is fast enough and doesn't need to be optimized to run just once, right?

File: src/wp-includes/blocks.php (hunk 3 above)
Comment (luisherranz) [reply src/wp-includes/blocks.php:1560]: Again, passing `$blocks` may depend on [how fast is `parse_blocks( $content )`](https://github.com/WordPress/wordpress-develop/pull/5023#issuecomment-1683610594) and if it makes sense to try to avoid doing it more than once.

Comment (dmsnell) [reply src/wp-includes/blocks.php:1560]: to make sure my comment was clear, I did consider what extra parameters we might want to pass but my personal thought when I did that was that none really belong.

> passing $blocks may depend on https://github.com/WordPress/wordpress-develop/pull/5023#issuecomment-1683610594 and if it makes sense to try to avoid doing it more than once.

are you suggesting that we re-parse the output and pass it along in the filter? this is another thing that any filter could compute on its own if it needs or wants to. there's the option of passing the original parsed blocks, but again, given that we don't have a good way to associate regions of output with the input blocks, it sets of my hesitancy.

do we have a compelling use-case we can think of to pass `$blocks` to the `do_blocks_post_render` function?

Comment (luisherranz) [reply src/wp-includes/blocks.php:1560]: > are you suggesting that we re-parse the output and pass it along in the filter? 

On the contrary. I was simply mentioning that if we do not pass `$blocks` to `do_block_post_render`, plugins that need it can access them by re-parsing, but that means that parsing would be done two (or more) times. And I don't know if that is concerning in terms of performance or not.

> Do we have a compelling use-case we can think of to pass $blocks to the do_blocks_post_render function?

Not that I can think of.

---

Ok, so what about this?

- `$blocks = apply_filters( 'do_blocks_pre_render', $blocks, $content );`
- `$output = apply_filters( 'do_blocks_post_render', $output, $content );`

The use case for `do_blocks_post_render` is to be able to move [this custom logic](https://github.com/WordPress/wordpress-develop/blob/5abfc9c4a2c81b82bdb489211db2493517e5b26b/src/wp-includes/blocks.php#L1196-L1201) outside of the `do_blocks` function.

I don't know any use case for passing `$content` to `do_blocks_pre_render`, but if we're passing it to the `post` filter, I think it makes sense to me to pass it to the `pre` filter as well. Anyway, if you prefer, we can omit that one until there's a clear use case.

cc: @gziolo @dmsnell

Comment (dmsnell) [reply src/wp-includes/blocks.php:1560]: My vote is almost always going to fall on the side of waiting until we have a need to provide a solution, which in this case is for omitting the second parameter. I'm not sure how someone is supposed to associate `$content` to `$blocks` or what the difference between `$content` and `$output` is, or again, how to associate them.

That's just my vote. As noted before though, the easier option is leaving them out and adding them once we need them, since they become additional arguments. It's definitely harder to remove them if we realize we shouldn't expose them.

Comment (luisherranz) [reply src/wp-includes/blocks.php:1560]: Fair enough, let's not add any additional arguments from the beginning then.

**Comment by luisherranz** [comment]
Ok, we won't add any additional arguments for the moment. I've now resolved this conversation: 

- https://github.com/WordPress/wordpress-develop/pull/5023#discussion_r1298134313

---

@gziolo, @dmsnell: Is there anything else to do here?

**Comment by gziolo** [comment]
> @gziolo, @dmsnell: Is there anything else to do here?

I'd like to make sure you saw and responded to the comment that @azaozz left on WordPress Trac: https://core.trac.wordpress.org/ticket/59131#comment:10.

**Comment by luisherranz** [comment]
> I'd like to make sure you saw and responded to the comment that azaozz left on WordPress Trac

Sorry, I somehow missed that. I answered now 🙂

**Comment by azaozz** [comment]
> Sorry, I somehow missed that.

Uh, sorry. Generally GH PRs are mostly for comments about code and/or implementation, Trac is more about architecture and why/why not. Both would work more or less so lets continue here :) 

> the_content filters only the post content, whereas blocks are nowadays also used on the theme (templates and template parts) and other places

Right. What worries me most about the proposed filters is the total lack of context. How would a plugin know when the blocks that are being rendered are for a post, a template, or some different purpose/place? That's why I asked for some code examples of how these filters are going to be used. Thinking it is pretty important to understand the intended usage, and to document it well in the docblocks.

> This is also potentially useful for things like word counters, advertising placement and so on.

Frankly I don't think so. All of these would work much better by using the DOM from JS. Processing HTML as string in PHP is a really bad way to do things, really. Always always has edge cases and errors, and couple lines of JS can totally change all of it on the front-end and make all the PHP processing completely useless. This conclusion comes from many years of "trying to do things with HTML", not just my experience but of many many developers :)

**Comment by luisherranz** [comment]
Thanks, Andrew.

> How would a plugin know when the blocks that are being rendered are for a post, a template, or some different purpose/place?

These filters are intended to be used precisely when you need to process the rendered blocks no matter where they come from, which is the case of the Interactivity API. That is why I think `do_blocks` is the best place to hook, because it's the function used in all situations where blocks are being processed.

> Frankly I don't think so. All of these would work much better by using the DOM from JS.

Yeah, you're probably right 🙂

But apart from that, this is still the only place we can use to inspect the render using a top-down approach, as I explained in the Trac comment:

> `render_blocks` filters are great, but they are limited in the order they run and the information they have: `render_block_data` and `render_block_context` are top-down and can only access the block instance information, not the final rendering. `render_block` has the final rendering, but it's bottom-up. The render is also difficult to inspect because each block contains not only its own render, but also the render of its inner blocks. The main need for these filters came from the Interactivity API directive processing, which needs to process things top-down and be able to separate the render of interactive blocks from their (non-interactive) inner blocks.

It's just not possible to do so before this [foreach loop](https://github.com/WordPress/wordpress-develop/pull/5023/files#diff-8c99af92e4ec0fdb307ddd9b42be1e1ef1efe4a9f31287c23f346244dddd1ce9R1191-R1193) has finished. And as the Interactivity API is powered by Preact in the frontend (which uses a top-down approach), we need to mimic that behavior or we would limit any directive that depends on a `Provider` component.

---

@mcsf, @youknowriad: you have a lot of experience with this, what do you think?

**Comment by dmsnell** [comment]
There are things we cannot do with the existing filters and which filtering the block parser doesn't give us a good hook for. There's no current flow in PHP which indicates when the blocks have finished rendering. I think these new filters are somewhat self-evident in that regard, and I wish we had put them in earlier.

Are there any reasons to avoid putting these in? There have been multiple real needs for them over the years, and I understand the desire to do more work in JS where the costs are pushed to the site viewer, but I don't see that preference eliminating the need on the server to initialize filters before block rendering and follow-up with post-processing. I would be sad to see a personal preference block valuable and needed work - this happened with the spoiler/details block and it delayed that for two years to the detriment of the project.

I also hear the point about being able to filter the block parser but I would be extremely reluctant to suggest that; it's a very different semantic match, and since I imagine plugins wanting to perform some initialization before block parsing, it's very awkward for them all to replace the parser and I'm guessing that would be an easy place for things to break. Whereas I see the filterable parser primarily a way to transform _stored posts in some form_ into _the tree of block nodes_, `do_blocks_pre_render` has an inherent expectation that we don't have to mix block parsing and block transformation.

This does all get complicated and open up new risks, but that does seem to be WordPress' strength with its filters. What, specifically, is a reason to block this work?

File: src/wp-includes/blocks.php [hunk 5]
```diff
@@ -1174,12 +1174,33 @@ function parse_blocks( $content ) {
...
+	/**
+	 * Filter that allows plugins to inspect or manipulate the array of parsed
+	 * block objects. The parsed blocks form a hierarchy, rather than a flat
+	 * array, as blocks can be nested within each other during parsing.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array[] $blocks Array of parsed block objects.
+	 */
+	$blocks = apply_filters( 'do_blocks_pre_render', $blocks );
```
Comment (dmsnell) [inline src/wp-includes/blocks.php:1529]: It would be really nice if these contained examples for how they might be used. I find the inline examples to be really valuable since some IDEs show them right where the cursor is. VSCode isn't that pretty, but it does show it.

Screenshot 2023-09-19 at 10 39 03 AM

Moreover those examples can be a great reference on the project documentation. See [enable_wp_debug_mode_checks](https://developer.wordpress.org/reference/hooks/enable_wp_debug_mode_checks/) for instance.

File: src/wp-includes/blocks.php [hunk 6]
```diff
@@ -1174,12 +1174,33 @@ function parse_blocks( $content ) {
  */
 function do_blocks( $content ) {
 	$blocks = parse_blocks( $content );
+
+	/**
+	 * Filter that allows plugins to inspect or manipulate the array of parsed
+	 * block objects. The parsed blocks form a hierarchy, rather than a flat
```
Comment (dmsnell) [inline src/wp-includes/blocks.php]: The note about the shape could be a great item in the description, where the summary remains only one or two lines long. It's essential that it allows manipulating the array of parsed block objects/nodes, but it's helpful extra information to remind us that the blocks are a tree.

File: src/wp-includes/blocks.php [hunk 7]
```diff
@@ -1174,12 +1174,33 @@ function parse_blocks( $content ) {
  */
 function do_blocks( $content ) {
 	$blocks = parse_blocks( $content );
+
+	/**
+	 * Filter that allows plugins to inspect or manipulate the array of parsed
```
Comment (dmsnell) [inline src/wp-includes/blocks.php]: It could be argued that it's implicit in the filter name, but we don't mention that this occurs before rendering. We could make that association stronger.

```php
/**
 * Filter that allows inspection and modification of the
 * tree of raw parsed blocks before they are rendered.
 *
```

File: src/wp-includes/blocks.php [hunk 8]
```diff
@@ -1174,12 +1174,33 @@ function parse_blocks( $content ) {
...
+	$blocks = apply_filters( 'do_blocks_pre_render', $blocks );
+
 	$output = '';
 
 	foreach ( $blocks as $block ) {
 		$output .= render_block( $block );
 	}
 
+	/**
+	 * Filter to allow plugins to inspect or manipulate the final HTML output.
```
Comment (dmsnell) [inline src/wp-includes/blocks.php]: here we use "filter to allow" but above we use "filter that allows" - it would be best if we harmonized these, unless there's a real difference in that language that's important.

Comment (dmsnell) [inline src/wp-includes/blocks.php]: as above, I think it could be useful to provide a tiny bit of extra context here despite the filter name.

```php
/**
 * Filter that allows processing the final HTML output from block rendering.
 *
 * This hook provides a post-processing step where it's possible to perform
 * cleanup or finish what was started during the rendering pass. This filter
 * runs only after all of the blocks have been rendered and there are no more
 * block objects available; only HTML.
 *
```

File: src/wp-includes/blocks.php [hunk 9]
```diff
@@ -1174,12 +1174,33 @@ function parse_blocks( $content ) {
...
+	 *
+	 * @since 6.4.0
```
Comment (dmsnell) [inline src/wp-includes/blocks.php:1554]: similar note to the first filter: this could go even further with a good example. for instance, a word-counter.

```php
 /**
  * …
  *
  * Example:
  *
  *     add_filter( 'do_blocks_post_render', function ( $html ) {
  *         $counter = sprintf( _( 'This post contains %1$s words.', count_words( $html ) );
  *         return "<p>{$counter}</p>{$html}";
  *     }, 10, 1 );
  *
  * …
```

File: tests/phpunit/tests/blocks/renderHooks.php [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+ * @group blocks
+ */
+class Tests_Blocks_Render_Hooks extends WP_UnitTestCase {
+	/**
+	 * @ticket 59131
+	 */
+	public function test_do_blocks_pre_render_filter() {
+		$remove_blocks_containing_wordpress = static function( $blocks ) {
+			foreach ( $blocks as $index => $block ) {
+				if ( strpos( $block['innerHTML'], 'WordPress' ) !== false ) {
```
Comment (dmsnell) [inline tests/phpunit/tests/blocks/renderHooks.php]: this is fine, but people are replacing these patterns with `str_contains( $block['innerHTML'], 'WordPress' )`

it's polyfilled in WordPress so shouldn't be a problem with PHP support I believe

File: tests/phpunit/tests/blocks/renderHooks.php [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+			<!-- wp:core/paragraph --><p>WordPress</p><!-- /wp:core/paragraph -->
+		';
+
+		add_filter( 'do_blocks_pre_render', $remove_blocks_containing_wordpress );
+
+		$result = do_blocks( $block_content );
+
+		remove_filter( 'do_blocks_pre_render', $remove_blocks_containing_wordpress );
+
+		$this->assertSame( '<p>Hello</p>', trim( $result ) );
```
Comment (dmsnell) [inline tests/phpunit/tests/blocks/renderHooks.php]: if the paragraph renders differently then this test might fail when it shouldn't. if we wanted to make the tests more resilient, we could separate the block content and run one manually without the excluded blocks, that way it would be solely testing the filter and not also the block rendering code.

```php
$hello_block = '<!-- wp:core/paragraph --><p>Hello</p><!-- /wp:core/paragraph -->';
$press_block = '<!-- wp:core/paragraph --><p>WordPress</p><!-- /wp:core/paragraph -->';

add_filter( 'do_blocks_pre_render', $remove_blocks_containing_wordpress );
$filtered_output = do_blocks( $hello_block . $press_block );
remove_filter( 'do_blocks_pre_render', $remove_blocks_containing_wordpress );

$wanted_output = do_blocks( $hello_block );
$this->assertSame( $wanted_output, $filtered_output, "Did not exclude the intended blocks before rendering." );
```

**Comment by youknowriad** [comment]
I've spent some time discussing this PR with @luisherranz and all the potential alternatives. I think this is the right call, these filters seem harmless and easy to maintain forward compatibility.

File: src/wp-includes/blocks.php (hunk 7 above)
Comment (costdev) [inline src/wp-includes/blocks.php]: Docblocks should always begin with a third-person singular verb, so "Filters" rather than "Filter". See [the handbook entry](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#language).

For example, the summary could then become:
```php
/**
 * Filters the tree of raw parsed blocks before they are rendered.
 *
 * @since...
 */
```

Comment (luisherranz) [reply src/wp-includes/blocks.php]: But isn't this referring to the filter itself, and not the action of filtering?

By the way, I was just renaming them to this, following [Dennis request](https://github.com/WordPress/wordpress-develop/pull/5023#pullrequestreview-1633871328):

- `Filter to allow inspection and modification of the tree of raw parsed blocks before they are rendered.`
- `Filter to allow processing the final HTML output from block rendering.`

File: src/wp-includes/blocks.php (hunk 8 above)
Comment (costdev) [inline src/wp-includes/blocks.php]: Similar to my previous comment, this could become:

```php
/**
 * Filters the final HTML output from block rendering.
 *
 * @since...
 */
```

File: src/wp-includes/blocks.php (hunk 7 above)
Comment (costdev) [reply src/wp-includes/blocks.php]: Core uses the same language for functions, methods, filters and actions, in that it describes **what it does**, rather than **what it is**.

File: src/wp-includes/blocks.php (hunk 8 above)
Comment (luisherranz) [reply src/wp-includes/blocks.php]: Oh, ok. I got you now. I'll replace the reference to the filter with the filter verb 👍

File: src/wp-includes/blocks.php (hunk 7 above)
Comment (costdev) [reply src/wp-includes/blocks.php]: You can see this example in the handbook entry I linked above:

> **Filters:** What is being filtered?
> Good: Filters the post content.
> Bad: Lets you edit the post content that is output in the post template.

Comment (luisherranz) [reply src/wp-includes/blocks.php]: [Nevermind](https://github.com/WordPress/wordpress-develop/pull/5023#discussion_r1340405194) 👍

File: src/wp-includes/blocks.php (hunk 8 above)
Comment (luisherranz) [reply src/wp-includes/blocks.php]: Done in https://github.com/WordPress/wordpress-develop/pull/5023/commits/7df954bf3eb0f317681e52e7cbaca6c551244244.

File: src/wp-includes/blocks.php (hunk 7 above)
Comment (luisherranz) [reply src/wp-includes/blocks.php]: Done in https://github.com/WordPress/wordpress-develop/pull/5023/commits/7df954bf3eb0f317681e52e7cbaca6c551244244.

File: src/wp-includes/blocks.php (hunk 9 above)
Comment (luisherranz) [reply src/wp-includes/blocks.php:1554]: Done in https://github.com/WordPress/wordpress-develop/pull/5023/commits/7df954bf3eb0f317681e52e7cbaca6c551244244.

File: src/wp-includes/blocks.php (hunk 8 above)
Comment (luisherranz) [reply src/wp-includes/blocks.php]: Done in https://github.com/WordPress/wordpress-develop/pull/5023/commits/7df954bf3eb0f317681e52e7cbaca6c551244244.

File: src/wp-includes/blocks.php (hunk 7 above)
Comment (luisherranz) [reply src/wp-includes/blocks.php]: Done in https://github.com/WordPress/wordpress-develop/pull/5023/commits/7df954bf3eb0f317681e52e7cbaca6c551244244.

File: src/wp-includes/blocks.php (hunk 6 above)
Comment (luisherranz) [reply src/wp-includes/blocks.php]: Done in https://github.com/WordPress/wordpress-develop/pull/5023/commits/7df954bf3eb0f317681e52e7cbaca6c551244244.

File: src/wp-includes/blocks.php (hunk 5 above)
Comment (luisherranz) [reply src/wp-includes/blocks.php:1529]: Done in https://github.com/WordPress/wordpress-develop/pull/5023/commits/7df954bf3eb0f317681e52e7cbaca6c551244244.

File: src/wp-includes/blocks.php (hunk 8 above)
Comment (luisherranz) [reply src/wp-includes/blocks.php]: I switched to "Filters the final HTML..." as requested by @costdev in https://github.com/WordPress/wordpress-develop/pull/5023#pullrequestreview-1649250659

File: tests/phpunit/tests/blocks/renderHooks.php (hunk 11 above)
Comment (luisherranz) [reply tests/phpunit/tests/blocks/renderHooks.php]: Done in https://github.com/WordPress/wordpress-develop/pull/5023/commits/453a3733cb464d4eb8dce2acf9550ca83138e9af.

File: tests/phpunit/tests/blocks/renderHooks.php (hunk 10 above)
Comment (luisherranz) [reply tests/phpunit/tests/blocks/renderHooks.php]: Done in https://github.com/WordPress/wordpress-develop/pull/5023/commits/453a3733cb464d4eb8dce2acf9550ca83138e9af.

**Comment by luisherranz** [comment]
Thanks for your reviews, @dmsnell  and @costdev. I've now addressed all the feedback and rebased the PR.

File: src/wp-includes/blocks.php [hunk 12]
```diff
@@ -1502,12 +1502,61 @@ function parse_blocks( $content ) {
...
+	 * Filters the tree of raw parsed blocks before they are rendered.
+	 *
+	 * The parsed blocks form a hierarchy, rather than a flat array, as blocks can
+	 * be nested within each other. Also, take into account that calls to
+	 * `do_blocks` could be nested if there are blocks that use `do_blocks` in
+	 * their own rendering.
+	 *
+	 * Example:
+	 *
+	 *     add_filter( 'do_blocks_pre_render', function ( $blocks ) {
```
Comment (peterwilsoncc) [inline src/wp-includes/blocks.php:1516]: It's not urgent given the discussion in Slack this-morning, but prior to 6.5 are you able to run this through the docs parser used by developer.wordpress.org to make sure the code example renders as expected.

There are instructions for setting up devhub at https://make.wordpress.org/docs/handbook/devhub/ -- it's been a while so I am not sure how up-to-date they are.

The WP-CLI command takes ages to run for the full code base, so  I suggest limiting it to this file only

File: src/wp-includes/blocks.php (hunk 7 above)
Comment (dmsnell) [reply src/wp-includes/blocks.php]: My bad. I was using Core's own code as inspiration - a terrible flaw 😆

File: src/wp-includes/blocks.php (hunk 12 above)
Comment (dmsnell) [reply src/wp-includes/blocks.php:1516]: > it's been a while so I am not sure how up-to-date they are.

I spent some time trying to get it working and had a lot of trouble, finally gave up.

initially it broke with `yarn@4+`, had trouble with version mismatches between `wp-cli` and `composer` and their configs in the repo, had similar trouble building the developer theme, and its parent theme, had issues with PHP8 (understandable), and then once I got it building things, it seemed like it created a separate page for every docblock but I couldn't find any linked and organized pages as are on the production page.

if someone with experience and knowledge could review the README I'd be super grateful because I've really wanted to be able to preview my docs changes, and I don't like having to wait for releases to go make changes.

Comment (peterwilsoncc) [reply src/wp-includes/blocks.php:1516]: > if someone with experience and knowledge could review the README I'd be super grateful because I've really wanted to be able to preview my docs changes, and I don't like having to wait for releases to go make changes.

I'll see if I can figure out the steps I took. It has been a while since I've needed to test docblocks.

**Comment by gziolo** [comment]
This one might not be necessary anymore after server side directives processing for Interactivity API landed with https://github.com/WordPress/wordpress-develop/commit/6917cac581da18022f71f2381c0c429d0a807052.

**Comment by luisherranz** [comment]
Feel free to remove the WP 6.5 label.
